src/System.Linq/tests/ToArrayToListToDictionaryTests.cs

```diff
@@ -0,0 +1,469 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+
+namespace System.Linq.Tests
+{
+    public class ToArrayToListToDictionaryTests
+    {
+        #region =============  Helper classes =============
+
+        private class EnumerableCollectionTest<T> : IEnumerable<T>
+        {
+            public T[] Items = new T[0];
+            public EnumerableCollectionTest(T[] items) { Items = items; }
+
+            public IEnumerator<T> GetEnumerator() { return ((IEnumerable<T>)Items).GetEnumerator(); }
+            IEnumerator IEnumerable.GetEnumerator() { return Items.GetEnumerator(); }
+        }
+        private class ReadOnlyCollectionTest<T> : IReadOnlyCollection<T>
+        {
+            public T[] Items = new T[0];
+            public int CountTouched = 0;
+            public ReadOnlyCollectionTest(T[] items) { Items = items; }
+
+            public int Count { get { CountTouched++; return Items.Length; } }
+            public IEnumerator<T> GetEnumerator() { return ((IEnumerable<T>)Items).GetEnumerator(); }
+            IEnumerator IEnumerable.GetEnumerator() { return Items.GetEnumerator(); }
+        }
+        private class CollectionTest<T> : ICollection<T>
+        {
+            public T[] Items = new T[0];
+            public int CountTouched = 0;
+            public int CopyToTouched = 0;
+            public CollectionTest(T[] items) { Items = items; }
+
+            public virtual int Count { get { CountTouched++; return Items.Length; } }
+            public bool IsReadOnly { get { return false; } }
+            public void Add(T item) { throw new NotImplementedException(); }
+            public void Clear() { throw new NotImplementedException(); }
+            public bool Contains(T item) { return Items.Contains(item); }
+            public bool Remove(T item) { throw new NotImplementedException(); }
+            public void CopyTo(T[] array, int arrayIndex) { CopyToTouched++; Items.CopyTo(array, arrayIndex); }
+            public IEnumerator<T> GetEnumerator() { return ((IEnumerable<T>)Items).GetEnumerator(); }
+            IEnumerator IEnumerable.GetEnumerator() { return Items.GetEnumerator(); }
+        }
+
+        private class LargeSequenceTest: IEnumerable<byte>
+        {
+            public long MaxSize = 2 * (long)int.MaxValue;
+            public IEnumerator<byte> GetEnumerator()
+            {
+                for (long i = 0; i < MaxSize; i++) yield return (byte)1;
+            }
+            IEnumerator IEnumerable.GetEnumerator() { return this.GetEnumerator(); }
+        }
+
+        /// <summary>
+        /// Emulation of async collection change
+        /// </summary>
+        private class AsyncCollectionTest : CollectionTest<int>
```

Comment:
This class helps to verify that ```ToArray``` and ```ToList``` can work with Concurrent collections. It adds a new element to the sequence each time the ```Count``` property read.
Probably you are right: 'Async' is not a good naming. Any suggestion how to rename it?